File: objectivec/day18_part2_2015.m
#import <Foundation/Foundation.h>

const int gridSize = 100;
const int steps = 100;

@interface Grid : NSObject {
    NSMutableArray *_grid;
}

- (instancetype)initWithContentsOfFile:(NSString *)path;
- (void)step;
- (int)countLightsOn;

@end

@implementation Grid

- (instancetype)initWithContentsOfFile:(NSString *)path {
    self = [super init];
    if (self) {
        NSError *error;
        NSString *contents = [NSString stringWithContentsOfFile:path encoding:NSUTF8StringEncoding error:&error];
        if (error) {
            NSLog(@"Error reading file: %@", error);
            return nil;
        }
        NSArray *lines = [contents componentsSeparatedByCharactersInSet:[NSCharacterSet newlineCharacterSet]];
        _grid = [NSMutableArray arrayWithCapacity:gridSize];
        for (int y = 0; y < gridSize; y++) {
            NSMutableArray *row = [NSMutableArray arrayWithCapacity:gridSize];
            NSString *line = lines[y];
            for (int x = 0; x < gridSize; x++) {
                [row addObject:@([line characterAtIndex:x] == '#')];
            }
            [_grid addObject:row];
        }
        [_grid[0] replaceObjectAtIndex:0 withObject:@YES];
        [_grid[0] replaceObjectAtIndex:gridSize-1 withObject:@YES];
        [_grid[gridSize-1] replaceObjectAtIndex:0 withObject:@YES];
        [_grid[gridSize-1] replaceObjectAtIndex:gridSize-1 withObject:@YES];
    }
    return self;
}

- (void)step {
    NSMutableArray *newGrid = [NSMutableArray arrayWithCapacity:gridSize];
    for (int y = 0; y < gridSize; y++) {
        NSMutableArray *row = [NSMutableArray arrayWithCapacity:gridSize];
        for (int x = 0; x < gridSize; x++) {
            int onNeighbors = 0;
            for (int dx = -1; dx <= 1; dx++) {
                for (int dy = -1; dy <= 1; dy++) {
                    if (dx == 0 && dy == 0) continue;
                    int nx = x + dx;
                    int ny = y + dy;
                    if (nx >= 0 && nx < gridSize && ny >= 0 && ny < gridSize && [_grid[ny][nx] boolValue]) {
                        onNeighbors++;
                    }
                }
            }
            if ([_grid[y][x] boolValue]) {
                [row addObject:@(onNeighbors == 2 || onNeighbors == 3)];
            } else {
                [row addObject:@(onNeighbors == 3)];
            }
        }
        [newGrid addObject:row];
    }
    _grid = newGrid;
    [_grid[0] replaceObjectAtIndex:0 withObject:@YES];
    [_grid[0] replaceObjectAtIndex:gridSize-1 withObject:@YES];
    [_grid[gridSize-1] replaceObjectAtIndex:0 withObject:@YES];
    [_grid[gridSize-1] replaceObjectAtIndex:gridSize-1 withObject:@YES];
}

- (int)countLightsOn {
    int count = 0;
    for (int y = 0; y < gridSize; y++) {
        for (int x = 0; x < gridSize; x++) {
            if ([_grid[y][x] boolValue]) count++;
        }
    }
    return count;
}

@end

int main(int argc, char *argv[]) {
    @autoreleasepool {
        Grid *grid = [[Grid alloc] initWithContentsOfFile:@"input.txt"];
        for (int i = 0; i < steps; i++) {
            [grid step];
        }
        NSLog(@"%d", [grid countLightsOn]);
    }
    return 0;
}

File: haskell/day18_part2_2015.hs
import System.IO

main :: IO ()
main = do
    contents <- readFile "input.txt"
    let grid = map (map (== '#')) $ lines contents
    let gridSize = length grid
    let fixedCorners = [(0, 0), (0, gridSize - 1), (gridSize - 1, 0), (gridSize - 1, gridSize - 1)]
    let finalGrid = iterate (step gridSize fixedCorners) (setFixedCorners grid fixedCorners) !! 100
    print $ countLightsOn finalGrid

step :: Int -> [(Int, Int)] -> [[Bool]] -> [[Bool]]
step gridSize fixedCorners grid = setFixedCorners newGrid fixedCorners
  where
    newGrid = [[nextState x y | y <- [0..gridSize-1]] | x <- [0..gridSize-1]]
    nextState x y = let onNeighbors = countOnNeighbors gridSize grid x y
                    in if grid !! x !! y
                       then onNeighbors == 2 || onNeighbors == 3
                       else onNeighbors == 3

countOnNeighbors :: Int -> [[Bool]] -> Int -> Int -> Int
countOnNeighbors gridSize grid x y = length [() | dx <- [-1..1], dy <- [-1..1], (dx, dy) /= (0, 0),
                                                   let nx = x + dx, let ny = y + dy,
                                                   nx >= 0, nx < gridSize, ny >= 0, ny < gridSize,
                                                   grid !! nx !! ny]

setFixedCorners :: [[Bool]] -> [(Int, Int)] -> [[Bool]]
setFixedCorners grid fixedCorners = foldl (\g (x, y) -> updateGrid g x y True) grid fixedCorners

updateGrid :: [[Bool]] -> Int -> Int -> Bool -> [[Bool]]
updateGrid grid x y val = take x grid ++ [take y (grid !! x) ++ [val] ++ drop (y + 1) (grid !! x)] ++ drop (x + 1) grid

countLightsOn :: [[Bool]] -> Int
countLightsOn grid = sum $ map (length . filter id) grid